File: src/javax/microedition/lcdui/Graphics.java
package javax.microedition.lcdui;

import java.awt.AlphaComposite;
import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.awt.RenderingHints;

public class Graphics {
  public static final int HCENTER = 1;
  public static final int VCENTER = 2;
  public static final int LEFT = 4;
  public static final int RIGHT = 8;
  public static final int TOP = 16;
  public static final int BOTTOM = 32;
  public static final int BASELINE = 64;

  public final Graphics2D g2d;
  private final boolean disposeG2d;
  private Font font;

  public Graphics(Graphics2D g2d, boolean disposeG2d) {
    this.g2d = g2d;
    this.disposeG2d = disposeG2d;
    this.font = Font.getDefaultFont();
    g2d.setFont(font.getAwtFont());
    g2d.setComposite(AlphaComposite.SrcOver);
    g2d.setRenderingHint(
        RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_NEAREST_NEIGHBOR);
    g2d.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_SPEED);
    g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_OFF);
  }

  public void dispose() {
    if (disposeG2d) {
      g2d.dispose();
    }
  }

  public void setColor(int rgb) {
    g2d.setColor(new Color((rgb >> 16) & 0xFF, (rgb >> 8) & 0xFF, rgb & 0xFF));
  }

  /** MIDP-style RGB components 0–255. */
  public void setColor(int r, int g, int b) {
    g2d.setColor(new Color(r & 0xFF, g & 0xFF, b & 0xFF));
  }

  public int getColor() {
    Color c = g2d.getColor();
    return (c.getRed() << 16) | (c.getGreen() << 8) | c.getBlue();
  }

  public void setFont(Font f) {
    this.font = f;
    g2d.setFont(f.getAwtFont());
  }

  public Font getFont() {
    return font;
  }

  private Rectangle clipBounds() {
    Rectangle r = g2d.getClipBounds();
    if (r == null) {
      return new Rectangle(0, 0, GameDimensions.getWidth(), GameDimensions.getHeight());
    }
    return r;
  }

  public int getClipX() {
    return clipBounds().x;
  }

  public int getClipY() {
    return clipBounds().y;
  }

  public int getClipWidth() {
    return clipBounds().width;
  }

  public int getClipHeight() {
    return clipBounds().height;
  }

  public void setClip(int x, int y, int width, int height) {
    g2d.setClip(x, y, width, height);
  }

  public void clipRect(int x, int y, int width, int height) {
    g2d.clipRect(x, y, width, height);
  }

  public void translate(int dx, int dy) {
    g2d.translate(dx, dy);
  }

  /** Current translation X from the graphics transform (MIDP-style). */
  public int getTranslateX() {
    return (int) g2d.getTransform().getTranslateX();
  }

  /** Current translation Y from the graphics transform (MIDP-style). */
  public int getTranslateY() {
    return (int) g2d.getTransform().getTranslateY();
  }

  public void fillRect(int x, int y, int w, int h) {
    g2d.fillRect(x, y, w, h);
  }

  public void fillArc(int x, int y, int width, int height, int startAngle, int arcAngle) {
    g2d.fillArc(x, y, width, height, startAngle, arcAngle);
  }

  public void drawRect(int x, int y, int w, int h) {
    g2d.drawRect(x, y, w, h);
  }

  public void fillRoundRect(int x, int y, int w, int h, int aw, int ah) {
    g2d.fillRoundRect(x, y, w, h, aw, ah);
  }

  public void drawRoundRect(int x, int y, int w, int h, int aw, int ah) {
    g2d.drawRoundRect(x, y, w, h, aw, ah);
  }

  public void drawLine(int x1, int y1, int x2, int y2) {
    g2d.drawLine(x1, y1, x2, y2);
  }

  public void drawImage(Image img, int x, int y, int anchor) {
    int w = img.getWidth();
    int h = img.getHeight();
    int[] tl = anchorTopLeft(x, y, w, h, anchor);
    if (img.cropped) {
      g2d.drawImage(img.buf, tl[0], tl[1], tl[0] + w, tl[1] + h,
          img.cropX, img.cropY, img.cropX + w, img.cropY + h, null);
    } else {
      g2d.drawImage(img.buf, tl[0], tl[1], null);
    }
  }

  public static int[] anchorTopLeft(int x, int y, int w, int h, int anchor) {
    int left = x;
    int top = y;
    if ((anchor & LEFT) != 0) {
      left = x;
    } else if ((anchor & HCENTER) != 0) {
      left = x - w / 2;
    } else if ((anchor & RIGHT) != 0) {
      left = x - w;
    }
    if ((anchor & TOP) != 0) {
      top = y;
    } else if ((anchor & VCENTER) != 0) {
      top = y - h / 2;
    } else if ((anchor & BOTTOM) != 0) {
      top = y - h;
    } else if ((anchor & BASELINE) != 0) {
      top = y;
    }
    return new int[] {left, top};
  }

  public void drawString(String str, int x, int y, int anchor) {
    if (str == null) {
      return;
    }
    java.awt.FontMetrics fm = g2d.getFontMetrics();
    int sw = fm.stringWidth(str);
    int sh = fm.getHeight();
    int ascent = fm.getAscent();
    int bx = x;
    int by = y;
    if ((anchor & LEFT) != 0) {
      bx = x;
    } else if ((anchor & HCENTER) != 0) {
      bx = x - sw / 2;
    } else if ((anchor & RIGHT) != 0) {
      bx = x - sw;
    }
    if ((anchor & TOP) != 0) {
      by = y + ascent;
    } else if ((anchor & VCENTER) != 0) {
      by = y + ascent - sh / 2;
    } else if ((anchor & BOTTOM) != 0) {
      by = y - (sh - ascent);
    } else if ((anchor & BASELINE) != 0) {
      by = y;
    } else {
      by = y + ascent;
    }
    g2d.drawString(str, bx, by);
  }

  public void drawSubstring(String str, int offset, int len, int x, int y, int anchor) {
    if (str == null || offset < 0 || len < 0 || offset + len > str.length()) {
      return;
    }
    drawString(str.substring(offset, offset + len), x, y, anchor);
  }

  public void fillTriangle(int x1, int y1, int x2, int y2, int x3, int y3) {
    int[] xs = {x1, x2, x3};
    int[] ys = {y1, y2, y3};
    g2d.fillPolygon(xs, ys, 3);
  }
}

File: docs/swingjs/j2s/javax/microedition/lcdui/Graphics.js
(function(){var P$=Clazz.newPackage("javax.microedition.lcdui"),p$1={},I$=[[0,'javax.microedition.lcdui.Font','java.awt.AlphaComposite','java.awt.RenderingHints','java.awt.Color','java.awt.Rectangle','javax.microedition.lcdui.GameDimensions']],I$0=I$[0],$I$=function(i,n){return((i=(I$[i]||(I$[i]=Clazz.load(I$0[i])))),!n&&i.$load$&&Clazz.load(i,2),i)};
/*c*/var C$=Clazz.newClass(P$, "Graphics");

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);

C$.$fields$=[['Z',['disposeG2d'],'O',['g2d','java.awt.Graphics2D','font','javax.microedition.lcdui.Font']]]

Clazz.newMeth(C$, 'c$$java_awt_Graphics2D$Z',  function (g2d, disposeG2d) {
;C$.$init$.apply(this);
this.g2d=g2d;
this.disposeG2d=disposeG2d;
this.font=$I$(1).getDefaultFont$();
g2d.setFont$java_awt_Font(this.font.getAwtFont$());
g2d.setComposite$java_awt_Composite($I$(2).SrcOver);
g2d.setRenderingHint$java_awt_RenderingHints_Key$O($I$(3).KEY_INTERPOLATION, $I$(3).VALUE_INTERPOLATION_NEAREST_NEIGHBOR);
g2d.setRenderingHint$java_awt_RenderingHints_Key$O($I$(3).KEY_RENDERING, $I$(3).VALUE_RENDER_SPEED);
g2d.setRenderingHint$java_awt_RenderingHints_Key$O($I$(3).KEY_ANTIALIASING, $I$(3).VALUE_ANTIALIAS_OFF);
}, 1);

Clazz.newMeth(C$, 'dispose$',  function () {
if (this.disposeG2d) {
this.g2d.dispose$();
}});

Clazz.newMeth(C$, 'setColor$I',  function (rgb) {
this.g2d.setColor$java_awt_Color(Clazz.new_([(rgb >> 16) & 255, (rgb >> 8) & 255, rgb & 255],$I$(4,1).c$$I$I$I));
});

Clazz.newMeth(C$, 'setColor$I$I$I',  function (r, g, b) {
this.g2d.setColor$java_awt_Color(Clazz.new_($I$(4,1).c$$I$I$I,[r & 255, g & 255, b & 255]));
});

Clazz.newMeth(C$, 'getColor$',  function () {
var c=this.g2d.getColor$();
return (c.getRed$() << 16) | (c.getGreen$() << 8) | c.getBlue$() ;
});

Clazz.newMeth(C$, 'setFont$javax_microedition_lcdui_Font',  function (f) {
this.font=f;
this.g2d.setFont$java_awt_Font(f.getAwtFont$());
});

Clazz.newMeth(C$, 'getFont$',  function () {
return this.font;
});

Clazz.newMeth(C$, 'clipBounds',  function () {
var r=this.g2d.getClipBounds$();
if (r == null ) {
return Clazz.new_([0, 0, $I$(6).getWidth$(), $I$(6).getHeight$()],$I$(5,1).c$$I$I$I$I);
}return r;
}, p$1);

Clazz.newMeth(C$, 'getClipX$',  function () {
return p$1.clipBounds.apply(this, []).x;
});

Clazz.newMeth(C$, 'getClipY$',  function () {
return p$1.clipBounds.apply(this, []).y;
});

Clazz.newMeth(C$, 'getClipWidth$',  function () {
return p$1.clipBounds.apply(this, []).width;
});

Clazz.newMeth(C$, 'getClipHeight$',  function () {
return p$1.clipBounds.apply(this, []).height;
});

Clazz.newMeth(C$, 'setClip$I$I$I$I',  function (x, y, width, height) {
this.g2d.setClip$I$I$I$I(x, y, width, height);
});

Clazz.newMeth(C$, 'clipRect$I$I$I$I',  function (x, y, width, height) {
this.g2d.clipRect$I$I$I$I(x, y, width, height);
});

Clazz.newMeth(C$, 'translate$I$I',  function (dx, dy) {
this.g2d.translate$I$I(dx, dy);
});

Clazz.newMeth(C$, 'getTranslateX$',  function () {
return (this.g2d.getTransform$().getTranslateX$()|0);
});

Clazz.newMeth(C$, 'getTranslateY$',  function () {
return (this.g2d.getTransform$().getTranslateY$()|0);
});

Clazz.newMeth(C$, 'fillRect$I$I$I$I',  function (x, y, w, h) {
this.g2d.fillRect$I$I$I$I(x, y, w, h);
});

Clazz.newMeth(C$, 'fillArc$I$I$I$I$I$I',  function (x, y, width, height, startAngle, arcAngle) {
this.g2d.fillArc$I$I$I$I$I$I(x, y, width, height, startAngle, arcAngle);
});

Clazz.newMeth(C$, 'drawRect$I$I$I$I',  function (x, y, w, h) {
this.g2d.drawRect$I$I$I$I(x, y, w, h);
});

Clazz.newMeth(C$, 'fillRoundRect$I$I$I$I$I$I',  function (x, y, w, h, aw, ah) {
this.g2d.fillRoundRect$I$I$I$I$I$I(x, y, w, h, aw, ah);
});

Clazz.newMeth(C$, 'drawRoundRect$I$I$I$I$I$I',  function (x, y, w, h, aw, ah) {
this.g2d.drawRoundRect$I$I$I$I$I$I(x, y, w, h, aw, ah);
});

Clazz.newMeth(C$, 'drawLine$I$I$I$I',  function (x1, y1, x2, y2) {
this.g2d.drawLine$I$I$I$I(x1, y1, x2, y2);
});

Clazz.newMeth(C$, 'drawImage$javax_microedition_lcdui_Image$I$I$I',  function (img, x, y, anchor) {
var w=img.getWidth$();
var h=img.getHeight$();
var tl=C$.anchorTopLeft$I$I$I$I$I(x, y, w, h, anchor);
if (img.cropped) {
this.g2d.drawImage$java_awt_Image$I$I$I$I$I$I$I$I$java_awt_image_ImageObserver(img.buf, tl[0], tl[1], tl[0] + w, tl[1] + h, img.cropX, img.cropY, img.cropX + w, img.cropY + h, null);
} else {
this.g2d.drawImage$java_awt_Image$I$I$java_awt_image_ImageObserver(img.buf, tl[0], tl[1], null);
}});

Clazz.newMeth(C$, 'anchorTopLeft$I$I$I$I$I',  function (x, y, w, h, anchor) {
var left=x;
var top=y;
if ((anchor & 4) != 0) {
left=x;
} else if ((anchor & 1) != 0) {
left=x - (w/2|0);
} else if ((anchor & 8) != 0) {
left=x - w;
}if ((anchor & 16) != 0) {
top=y;
} else if ((anchor & 2) != 0) {
top=y - (h/2|0);
} else if ((anchor & 32) != 0) {
top=y - h;
} else if ((anchor & 64) != 0) {
top=y;
}return Clazz.array(Integer.TYPE, -1, [left, top]);
}, 1);

Clazz.newMeth(C$, 'drawString$S$I$I$I',  function (str, x, y, anchor) {
if (str == null ) {
return;
}var fm=this.g2d.getFontMetrics$();
var sw=fm.stringWidth$S(str);
var sh=fm.getHeight$();
var ascent=fm.getAscent$();
var bx=x;
var by=y;
if ((anchor & 4) != 0) {
bx=x;
} else if ((anchor & 1) != 0) {
bx=x - (sw/2|0);
} else if ((anchor & 8) != 0) {
bx=x - sw;
}if ((anchor & 16) != 0) {
by=y + ascent;
} else if ((anchor & 2) != 0) {
by=y + ascent - (sh/2|0);
} else if ((anchor & 32) != 0) {
by=y - (sh - ascent);
} else if ((anchor & 64) != 0) {
by=y;
} else {
by=y + ascent;
}this.g2d.drawString$S$I$I(str, bx, by);
});

Clazz.newMeth(C$, 'drawSubstring$S$I$I$I$I$I',  function (str, offset, len, x, y, anchor) {
if (str == null  || offset < 0  || len < 0  || offset + len > str.length$() ) {
return;
}this.drawString$S$I$I$I(str.substring$I$I(offset, offset + len), x, y, anchor);
});

Clazz.newMeth(C$, 'fillTriangle$I$I$I$I$I$I',  function (x1, y1, x2, y2, x3, y3) {
var xs=Clazz.array(Integer.TYPE, -1, [x1, x2, x3]);
var ys=Clazz.array(Integer.TYPE, -1, [y1, y2, y3]);
this.g2d.fillPolygon$IA$IA$I(xs, ys, 3);
});

Clazz.newMeth(C$);
})();
;Clazz.setTVer('3.3.1-v5');//Created 2026-06-25 15:29:03 Java2ScriptVisitor version 3.3.1-v5 net.sf.j2s.core.jar version 3.3.1-v5
